File: baserecycleradapter/src/main/java/com/gigigo/baserecycleradapter/debouncedlisteners/DebouncedOnClickListener.java
package com.gigigo.baserecycleradapter.debouncedlisteners;

import android.view.View;
import com.gigigo.baserecycleradapter.viewholder.BaseViewHolder;

public abstract class DebouncedOnClickListener implements BaseViewHolder.OnItemClickListener, DebouncedListener {

  private final DebouncedClickHandler debouncedClickHandler;

  protected DebouncedOnClickListener() {
    this.debouncedClickHandler = new DebouncedClickHandler(this);
  }

  protected DebouncedOnClickListener(long millisIntervalToAvoidDoubleClick) {
    this.debouncedClickHandler = new DebouncedClickHandler(this, millisIntervalToAvoidDoubleClick);
  }

  @Override public void onItemClick(int position, View view) {
    debouncedClickHandler.invokeDebouncedClick(position, view);
  }
}

File: baserecycleradapter/src/main/java/com/gigigo/baserecycleradapter/debouncedlisteners/DebouncedOnClickListener.kt
package com.gigigo.baserecycleradapter.debouncedlisteners

import android.view.View
import com.gigigo.baserecycleradapter.viewholder.OnItemClickListener

internal interface DebouncedListener {
    fun onDebouncedClick(v: View, position: Int): Boolean
}

internal abstract class DebouncedOnClickListener protected constructor(
    millisIntervalToAvoidDoubleClick: Long? = null
) : OnItemClickListener, DebouncedListener {

    @Suppress("LeakingThis")
    private val debouncedClickHandler: DebouncedClickHandler =
        DebouncedClickHandler(this, millisIntervalToAvoidDoubleClick)

    override fun invoke(position: Int, view: View) {
        debouncedClickHandler.invokeDebouncedClick(position, view)
    }
}
